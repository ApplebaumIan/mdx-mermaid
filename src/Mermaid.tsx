/**
 * Copyright (c) Samuel Wall.
 *
 * This source code is licensed under the MIT license found in the
 * license file in the root directory of this source tree.
 */

import React, { useEffect, ReactElement } from 'react'
import mermaid from 'mermaid'

import { Config } from './config.model'

/**
 * Properties for Mermaid component.
 */
export type MermaidProps = {
  /**
   * Mermaid diagram.
   */
  chart: string

  /**
   * Config to initialize mermaid with.
   */
  config?: Config
}

/**
 * Component to display Mermaid diagrams.
 *
 * @param param0 Diagram to display.
 * @param param1 Config.
 * @returns The component.
 */
export const Mermaid = ({ chart, config }: MermaidProps): ReactElement<MermaidProps> => {
  // init diagram once
  useEffect(() => {
    if (config && config.mermaid) {
      mermaid.initialize({ startOnLoad: true, ...config.mermaid })
    }
    mermaid.contentLoaded()
  }, [])
  return <div className="mermaid">{chart}</div>
}
