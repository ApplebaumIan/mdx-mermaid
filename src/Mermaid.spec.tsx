/**
 * @jest-environment jsdom
 */
/**
 * Copyright (c) Samuel Wall.
 *
 * This source code is licensed under the MIT license found in the
 * license file in the root directory of this source tree.
 */
import React from 'react'
import renderer from 'react-test-renderer'

import mermaid from 'mermaid'
import { Mermaid } from './Mermaid'

jest.mock('mermaid')

afterEach(() => {
  jest.clearAllMocks()
})

it('renders without diagram', () => {
  const component = renderer.create(<Mermaid chart={''} />)
  expect(component.toJSON()).toMatchSnapshot()
  expect(mermaid.initialize).toBeCalledTimes(0)
  expect(mermaid.contentLoaded).toBeCalledTimes(0)
  component.update()
  expect(mermaid.contentLoaded).toHaveBeenCalled()
  expect(mermaid.initialize).toBeCalledTimes(0)
  component.update()
  expect(mermaid.contentLoaded).toBeCalledTimes(1)
  expect(mermaid.initialize).toBeCalledTimes(0)
})

it('renders with diagram', () => {
  const component = renderer.create(<Mermaid chart={`graph TD;
      A-->B;
      A-->C;
      B-->D;
      C-->D;`} />)
  expect(component.toJSON()).toMatchSnapshot()
  expect(mermaid.initialize).toBeCalledTimes(0)
  expect(mermaid.contentLoaded).toBeCalledTimes(0)
  component.update()
  expect(mermaid.contentLoaded).toHaveBeenCalled()
  expect(mermaid.initialize).toBeCalledTimes(0)
  component.update()
  expect(mermaid.contentLoaded).toBeCalledTimes(1)
  expect(mermaid.initialize).toBeCalledTimes(0)
})

it('renders with config', () => {
  const component = renderer.create(<Mermaid chart={`graph TD;
      A-->B;
      A-->C;
      B-->D;
      C-->D;`} config={{}} />)
  expect(component.toJSON()).toMatchSnapshot()
  expect(mermaid.initialize).toBeCalledTimes(0)
  expect(mermaid.contentLoaded).toBeCalledTimes(0)
  component.update()
  expect(mermaid.contentLoaded).toHaveBeenCalled()
  expect(mermaid.initialize).toBeCalledTimes(0)
  component.update()
  expect(mermaid.contentLoaded).toBeCalledTimes(1)
  expect(mermaid.initialize).toBeCalledTimes(0)
})

it('renders with mermaid config', () => {
  const component = renderer.create(<Mermaid chart={`graph TD;
      A-->B;
      A-->C;
      B-->D;
      C-->D;`} config={{ mermaid: { theme: 'dark' } } } />)
  expect(component.toJSON()).toMatchSnapshot()
  expect(mermaid.initialize).toBeCalledTimes(0)
  expect(mermaid.contentLoaded).toBeCalledTimes(0)
  component.update()
  expect(mermaid.contentLoaded).toHaveBeenCalled()
  expect(mermaid.initialize).toHaveBeenNthCalledWith(1, { startOnLoad: true, theme: 'dark' })
  component.update()
  expect(mermaid.contentLoaded).toBeCalledTimes(1)
  expect(mermaid.initialize).toBeCalledTimes(1)
})
