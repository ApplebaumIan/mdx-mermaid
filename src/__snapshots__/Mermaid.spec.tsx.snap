// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with config 1`] = `
<div
  className="mermaid"
>
  graph TD;
      A--&gt;B;
      A--&gt;C;
      B--&gt;D;
      C--&gt;D;
</div>
`;

exports[`renders with diagram 1`] = `
<div
  className="mermaid"
>
  graph TD;
      A--&gt;B;
      A--&gt;C;
      B--&gt;D;
      C--&gt;D;
</div>
`;

exports[`renders with mermaid config 1`] = `
<div
  className="mermaid"
>
  graph TD;
      A--&gt;B;
      A--&gt;C;
      B--&gt;D;
      C--&gt;D;
</div>
`;

exports[`renders without diagram 1`] = `
<div
  className="mermaid"
>
  
</div>
`;
